File: cpp/src/array.cpp
#include <iostream>
#include <string>
#include <fstream>
#include "array.h"

using namespace std;

void push_index(Array *arr, int index, const string &key){
    string *newV = new string[arr->size + 1];
    int i = 0;
    while (i != index){
        newV[i] = arr->value[i];
        i++;
    }
    newV[i] = key;
    while (i < arr->size){
        newV[i+1] = arr->value[i];
        i++;
    }
    delete [] arr->value;
    arr->value = newV;
    arr->size++;
}

void push_back(Array *arr, const string &key){
    string *newV = new string[arr->size + 1];
    int i = 0;
    while (i < arr->size){
        newV[i] = arr->value[i];
        i++;
    }
    newV[i] = key;
    delete [] arr->value;
    arr->value = newV;
    arr->size++;
}

void pop_index(Array *arr, int index){
    if (index >= arr->size) return;
    string *newV =  new string[arr->size - 1];
    int i = 0;
    while (i != index){
        newV[i] = arr->value[i];
        i++;
    }
    while (i < arr->size - 1){
        newV[i] = arr->value[i+1];
        i++;
    }
    delete [] arr->value;
    arr->value = newV;
    arr->size--;
}

void replace(Array *arr, int index, const string& key){
    if (index < 0 or index >= arr->size) return;
    arr->value[index] = key;
}

void print(Array *arr){
    for (int i = 0; i < arr->size; i++){
        cout << arr->value[i] << " ";
    }
    cout << endl;
}

string get(Array *arr, int index){
    if (index < 0 or index >= arr->size) return "Out of range";
    return arr->value[index];
}

int length(Array *arr){
    return arr->size;
}

void saveToFile(Array *arr, const string &filename) {
    ofstream file(filename);
    if (!file.is_open()) {
        cout << "Error opening file for writing!" << endl;
        return;
    }
    
    for (int i = 0; i < arr->size; i++) {
        file << arr->value[i] << endl;
    }
    
    file.close();
}

void loadFromFile(Array *arr, const string &filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cout << "Error opening file for reading!" << endl;
        return;
    }
    
    clear(arr);
    string line;
    while (getline(file, line)) {
        push_back(arr, line);
    }
    
    file.close();
}

void clear(Array *arr) {
    delete[] arr->value;
    arr->size = 0;
    arr->value = new string[0];
}

File: go/array.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Array struct {
	size  int
	value []string
}

func NewArray() *Array {
	return &Array{
		size:  0,
		value: make([]string, 0),
	}
}

func pushIndex(arr *Array, index int, key string) {
	newV := make([]string, arr.size+1)

	i := 0
	for i != index {
		newV[i] = arr.value[i]
		i++
	}

	newV[i] = key

	for i < arr.size {
		newV[i+1] = arr.value[i]
		i++
	}

	arr.value = newV
	arr.size++
}

func pushBack(arr *Array, key string) {
	arr.value = append(arr.value, key)
	arr.size++
}

func popIndex(arr *Array, index int) {
	if index >= arr.size {
		return
	}

	newV := make([]string, 0, arr.size-1)
	newV = append(newV, arr.value[:index]...)
	newV = append(newV, arr.value[index+1:]...)

	arr.value = newV
	arr.size--
}

func replace(arr *Array, index int, key string) {
	if index < 0 || index >= arr.size {
		return
	}
	arr.value[index] = key
}

func printArray(arr *Array) {
	fmt.Println(strings.Join(arr.value, " "))
}

func get(arr *Array, index int) string {
	if index < 0 || index >= arr.size {
		return "Out of range"
	}
	return arr.value[index]
}

func length(arr *Array) int {
	return arr.size
}

func saveToFileArray(arr *Array, filename string) {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Println("Error opening file for writing:", err)
		return
	}
	defer file.Close()

	for _, val := range arr.value {
		_, err := file.WriteString(val + "\n")
		if err != nil {
			fmt.Println("Error writing to file:", err)
			return
		}
	}
}

func loadFromFileArray(arr *Array, filename string) {
	content, err := os.ReadFile(filename)
	if err != nil {
		fmt.Println("Error opening file for reading:", err)
		return
	}

	clearArray(arr)
	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		if line != "" {
			pushBack(arr, line)
		}
	}
}

func clearArray(arr *Array) {
	arr.value = make([]string, 0)
	arr.size = 0
}
